debezium-storage/debezium-storage-jdbc/src/main/java/io/debezium/storage/jdbc/RecoverableConnectionHolder.java

```diff
@@ -0,0 +1,130 @@
+        this.url = url;
+        this.user = user;
+        this.pwd = pwd;
+        this.waitRetryDelay = waitRetryDelay;
+        this.maxRetryCount = maxRetryCount;
+
+        createConnection();
+    }
+
+    private void createConnection() throws SQLException {
+        this.conn = DriverManager.getConnection(url, user, pwd);
+        this.conn.setAutoCommit(false);
+    }
+
+    @Override
+    public void close() throws SQLException {
+        conn.close();
+    }
+
+    public boolean isConnectionCreated() {
+        return conn != null;
+    }
+
+    public synchronized <T> T executeWithReconnect(ConnectionFunction<T> func, String name, boolean rollback)
+            throws SQLException, CustomException {
+        int attempt = 1;
+        while (true) {
+            if (conn == null) {
+                LOGGER.debug("Trying to reconnect (attempt {}).", attempt);
+                try {
+                    createConnection();
+                }
+                catch (SQLException e) {
+                    LOGGER.error("SQL Exception while trying to reconnect: " + e.getMessage(), e);
+                    conn = null;
+                    if (attempt >= maxRetryCount) {
+                        throw e;
+                    }
+                    attempt++;
+                    LOGGER.debug("Waiting for reconnect for {} ms.", waitRetryDelay);
+                    DelayStrategy delayStrategy = DelayStrategy.constant(waitRetryDelay);
+                    delayStrategy.sleepWhen(true);
+                    continue;
+                }
+            }
+            try {
+                return func.apply(conn);
+            }
+            catch (SQLRecoverableException e) {
+                LOGGER.warn("Attempt {} to call '{}' failed.", attempt, name, e);
+                if (rollback) {
+                    LOGGER.warn("'{}': doing rollback.", name);
+                    try {
+                        conn.rollback();
+                    }
+                    catch (SQLException ex) {
+                        // ignore
+                    }
+                }
+                try {
+                    conn.close();
+                }
+                catch (Exception ex) {
+                    // ignore
+                }
+                conn = null;
+
+            }
+            catch (SQLException e) {
+                LOGGER.warn("Call '{}' failed.", name, e);
+                if (rollback) {
+                    LOGGER.warn("'{}': doing rollback.", name);
+                    try {
+                        conn.rollback();
+                    }
+                    catch (SQLException ex) {
+                        // ignore
+                    }
+                }
+                throw e;
+            }
+        }
+    }
+
+    @FunctionalInterface
+    public interface ConnectionFunction<T> {
```

Comment:
I'd like to see TWO functional interfaces. One as is now and another that will be derived from `Consumer` for calls not returning any return value. There should be two `executeWithReconnect` each taking different parameter.

Reply:
No there is no need for code duplication, you can implement it via a single method. What is important is not to force the caller to return dummy `null` value if nothing is going to be returned.
So the new method can just delegate to the old one or there can be a thir `private` called by the new and old.

Reply:
that's where I'm not sure how to bridge between the 2 functional interfaces with different signature
Is this cast going to work?

```
public synchronized void executeWithRetry(ConnectionVoidFunction func, String name, boolean rollback)
            throws SQLException, RetriableException {
        executeWithRetry((ConnectionFunction<? extends Object>) func, name, rollback);
    }
```

I have 

```
    @FunctionalInterface
    public interface ConnectionFunction<T> {
        T apply(Connection conn) throws SQLException, RetriableException;
    }

    @FunctionalInterface
    public interface ConnectionVoidFunction {
        void apply(Connection conn) throws SQLException, RetriableException;
    }
```

Reply:
First, please mimic `java.util.function.Consumer<T>` interface for no return value.

The easiest implementation, the third private method accepts both function based and consumer based parameters. The inside of the method will be the same as the current one but will check which of those two parameters is not null and call appropriate function/consumer.

The two public methods will call the private one passing the parameter they got and `null` for the other one.

Reply:
thanks for guidance